### ISSUE TITLE
FID of Tiny-ImageNet or ImageNet 64x64

### ISSUE THREAD

**Issue by Yeez-lee** [FutureXiang/ddae#4]
Hi,

Thanks for your codes. But I have questions about FID when dataset is larger. If my dataset is either Tiny-ImageNet or ImageNet 64x64, how many images should I generate to calculate FID?  The exact number of Tiny-ImageNet or ImageNet 64x64 (larger than 50k)? And I should change the batch number (125) and 400 (125*400=50k) in **sample.py**, right?

BTW, I see other codes use **total_training_steps** instead of **epoch**. What is the relationship between these?

**Comment by FutureXiang** [comment]
Hi,

Thank you for your interest. 

1. The standard way to report FID is to (1) generate **50k** images and (2) compare them to the source dataset. For example, you may want to use 50k generated images and 1.28m ImageNet images to calculate FID on IN64x64. It is NOT necessary to keep the source set size (e.g., 50k Cifar / 100k Tiny-IN /1.28m IN) the same as the target set size (50k), because 50k is large enough.
    - However, for experimental purposes (e.g., checkpoint & hyper-parameter selection), you may want to monitor the FID using only **10k** images, which is more efficient.
    - Please note that the FID calculation module used in this repo (i.e., `pytorch-fid`) may NOT be suitable for very large dataset, because it extracts features and statistics for the source set repeatedly on each calculation. To do this more efficiently and elegantly, please check the [EDM repo](https://github.com/NVlabs/edm#calculating-fid).
2. I prefer using `num_epochs` because the total number of training images can be determined, given a specific dataset. Likewise, the [EDM training code](https://github.com/NVlabs/edm/blob/62072d2612c7da05165d6233d13d17d71f213fee/training/training_loop.py#L36) uses `total_kimg` to represent training duration. In contrast, `total_training_steps` is kind of meaningless in terms of representing the training budget because different implementations may use different `batch_size`.

**Comment by Yeez-lee** [comment]
Hi, 
Thank you for your reply. And I want to make the following comments.
(1) Even for the larger dataset (100k Tiny-IN /1.28m IN), only 50k generated images with source images  (100k or 1.28m) are enough for FID. Is this correct?
(2) In your codes, does `num_epochs` mean that **in each epoch, every data in the dataset is trained**? For CIFAR-10, if I have 1000 epochs, do I have **1000*50k** training images totally (like `total_kimg` in EDM)? If  `total_kimg` is smaller than the size of a dataset, does it mean that not every data is trained?

**Comment by Yeez-lee** [comment]
Thanks for your help. And I notice that your work uses unconditional models (DDPM or EDM). What if these models are conditional ones? DDAE (DiT-XL/2) is evaluated in an unconditional manner, but how about conditional models (DDPM or EDM)?

**Comment by FutureXiang** [comment]
The CFG models (which are joint 10% uncond + 90% cond models) yield worse representations than pure unconditional models, despite achieving SOTA generative FIDs.
- If we use `y=null` to retrieve features (i.e. in an unconditional manner), the features are still somewhat good, but the performance degrades (e.g. unconditional models reach ~90% K-NN acc on CIFAR-10, while CFG ones reach ~84%). I think (1) the insufficient 10% training of the unconditional version and (2) the joint parameterization limit the performance.
- If we pass non-null `y in [1...C]` label embeddings to retrieve features, these features are basically meaningless.
- An alternative way to use conditional models for classification is to probe the correct label in a zero-shot manner [arxiv1](https://arxiv.org/pdf/2303.15233.pdf) [arxiv2](https://arxiv.org/pdf/2303.16203.pdf), which is similar to CLIP. However, this way is somewhat inefficient because it relies on $T\times C$ forward passes to infer an answer.
